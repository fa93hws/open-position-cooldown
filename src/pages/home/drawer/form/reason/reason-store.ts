import type { ComponentType } from 'react';
import { makeObservable, observable, action, computed } from 'mobx';

import { nonEmpty } from '../validator';
import { InputStore } from '../input/input-store';
import { createInput, ExposedInputProps } from '../input/input';

export class ReasonStore {
  reasons: {
    Component: ComponentType<ExposedInputProps>;
    store: InputStore;
  }[] = [];

  shouldShowRemove = false;

  constructor() {
    makeObservable(this, {
      reasons: observable.deep,
      shouldShowRemove: observable.ref,
      addReason: action,
      removeReason: action,
      ReasonInputs: computed,
      setRemoveVisibility: action,
    });
  }

  get ReasonInputs(): ComponentType<ExposedInputProps>[] {
    return this.reasons.map((reason) => reason.Component);
  }

  get reasonStrings(): string[] {
    return this.reasons.map((reason) => reason.store.value);
  }

  get reasonErrors(): boolean[] {
    return this.reasons.map((reason) => reason.store.hasError);
  }

  addReason(): InputStore {
    const [Component, store] = createInput([nonEmpty]);
    this.reasons.push({ Component, store });
    return store;
  }

  removeReason(idx: number) {
    this.reasons.splice(idx, 1);
  }

  setRemoveVisibility(val: boolean) {
    this.shouldShowRemove = val;
  }
}
