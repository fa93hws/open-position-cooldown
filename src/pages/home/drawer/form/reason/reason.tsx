import * as React from 'react';
import {
  Box,
  Typography,
  IconButton,
  Switch,
  List,
  ListItem,
} from '@material-ui/core';
import { AddBox, Delete } from '@material-ui/icons';
import { makeStyles, withTheme, WithTheme } from '@material-ui/core/styles';
import { observer } from 'mobx-react';
import IndeterminateCheckBoxIcon from '@material-ui/icons/IndeterminateCheckBox';

import type { ExposedInputProps } from '../input/input';
import { ReasonStore } from './reason-store';

type ReasonInputItemProps = WithTheme & {
  ReasonInput: React.ComponentType<ExposedInputProps>;
  idx: number;
  shouldShowRemove: boolean;
  onRemoveClick(idx: number): void;
};

const useStyles = makeStyles((theme) => ({
  iconButton: {
    padding: 0,
    marginLeft: theme.spacing(1),
  },
  reasonList: {
    padding: 0,
  },
}));

const ReasonInputItem = React.memo(
  withTheme((props: ReasonInputItemProps) => {
    const onRemoveClick = React.useCallback(
      () => props.onRemoveClick(props.idx),
      [props.idx, props.onRemoveClick],
    );
    const styles = useStyles(props.theme);
    return (
      <ListItem dense disableGutters>
        <props.ReasonInput multiline={true} placeholder="可换行" />
        {props.shouldShowRemove && (
          <IconButton
            color="secondary"
            className={styles.iconButton}
            onClick={onRemoveClick}
            id={`remove-${props.idx}`}
          >
            <IndeterminateCheckBoxIcon />
          </IconButton>
        )}
      </ListItem>
    );
  }),
);

type ReasonSectionProps = WithTheme & {
  readonly ReasonsInput: React.ComponentType<ExposedInputProps>[];
  onAddClick(): void;
  shouldShowRemove: boolean;
  onSwitchChange(val: boolean): void;
  onRemoveClick(idx: number): void;
};

const useSwitchStyles = makeStyles({
  sizeSmall: {
    '& $switchBase': {
      padding: 0,
    },
  },
  root: {
    marginLeft: 'auto',
  },
  switchBase: {},
});

export const ReasonSection = React.memo(
  withTheme((props: ReasonSectionProps) => {
    const styles = useStyles(props.theme);
    const switchStyles = useSwitchStyles();
    const UncheckIcon = <Delete color="secondary" />;
    const CheckedIcon = <Delete />;
    const onSwitchChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
      props.onSwitchChange(ev.target.checked);
    };
    return (
      <Box>
        <Box display="flex" alignItems="center">
          <Typography variant="h6" component="h4">
            持有逻辑(至少3条)
          </Typography>
          <Switch
            id="switch"
            size="small"
            color="primary"
            checked={props.shouldShowRemove}
            checkedIcon={CheckedIcon}
            icon={UncheckIcon}
            classes={switchStyles}
            onChange={onSwitchChange}
          />
          <IconButton
            id="add-reason"
            color="primary"
            className={styles.iconButton}
            onClick={props.onAddClick}
          >
            <AddBox />
          </IconButton>
        </Box>
        <List>
          {props.ReasonsInput.map((ReasonInput, idx) => (
            <ReasonInputItem
              ReasonInput={ReasonInput}
              shouldShowRemove={props.shouldShowRemove}
              onRemoveClick={props.onRemoveClick}
              key={idx}
              idx={idx}
            />
          ))}
        </List>
      </Box>
    );
  }),
);

export function createReasonSection(): [React.ComponentType, ReasonStore] {
  const store = new ReasonStore();
  store.addReason();
  store.addReason();
  store.addReason();
  const onAddClick = () => store.addReason();
  const onRemoveClick = (idx: number) => store.removeReason(idx);
  const onSwitchChange = (val: boolean) => store.setRemoveVisibility(val);

  const Component = observer(() => (
    <ReasonSection
      ReasonsInput={store.ReasonInputs}
      onAddClick={onAddClick}
      shouldShowRemove={store.shouldShowRemove}
      onSwitchChange={onSwitchChange}
      onRemoveClick={onRemoveClick}
    />
  ));
  return [Component, store];
}
