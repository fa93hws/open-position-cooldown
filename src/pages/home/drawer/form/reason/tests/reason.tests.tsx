import * as React from 'react';
import { mount } from 'enzyme';

import { createReasonSection, ReasonSection } from '../reason';

describe('ReasonSection', () => {
  it('renders the section without any inputs', () => {
    expect(
      <ReasonSection
        ReasonsInput={[]}
        onAddClick={jest.fn()}
        shouldShowRemove={false}
        onSwitchChange={jest.fn()}
        onRemoveClick={jest.fn()}
      />,
    ).toMatchRenderedSnapshot();
  });

  it('renders the section with two input', () => {
    expect(
      <ReasonSection
        ReasonsInput={[() => <div>input 1</div>, () => <div>input 2</div>]}
        onAddClick={jest.fn()}
        shouldShowRemove={false}
        onSwitchChange={jest.fn()}
        onRemoveClick={jest.fn()}
      />,
    ).toMatchRenderedSnapshot();
  });

  it('renders the section with two input and remove icon', () => {
    expect(
      <ReasonSection
        ReasonsInput={[() => <div>input 1</div>, () => <div>input 2</div>]}
        onAddClick={jest.fn()}
        shouldShowRemove
        onSwitchChange={jest.fn()}
        onRemoveClick={jest.fn()}
      />,
    ).toMatchRenderedSnapshot();
  });

  it('show three reasons by default', () => {
    const [, store] = createReasonSection();
    expect(store.reasons.length).toEqual(3);
  });

  it('can display remove icon by switch', () => {
    const [Component, store] = createReasonSection();
    const element = mount(<Component />);
    element
      .find('input#switch')
      .simulate('change', { target: { checked: true } });
    expect(store.shouldShowRemove).toEqual(true);
  });

  it('adds an input when add is clicked', () => {
    const [Component, store] = createReasonSection();
    const element = mount(<Component />);
    element.find('button#add-reason').simulate('click');
    expect(store.reasons.length).toEqual(4);
  });

  it('remove the input at idx when remove is clicked', () => {
    const [Component, store] = createReasonSection();
    store.shouldShowRemove = true;
    const reasons = [...store.reasons];
    const element = mount(<Component />);
    element.find('button#remove-1').simulate('click');
    expect(store.reasons).toEqual([reasons[0], reasons[2]]);
  });
});
