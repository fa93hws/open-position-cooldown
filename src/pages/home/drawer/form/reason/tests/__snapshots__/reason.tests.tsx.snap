// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReasonSection renders the section with two input 1`] = `
<div
  class="MuiBox-root MuiBox-root-12"
>
  <div
    class="MuiBox-root MuiBox-root-13"
  >
    <h4
      class="MuiTypography-root MuiTypography-h6"
    >
      持有逻辑(至少3条)
    </h4>
    <span
      class="MuiSwitch-root makeStyles-root-4 MuiSwitch-sizeSmall makeStyles-sizeSmall-3"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiSwitch-switchBase makeStyles-switchBase-5 MuiSwitch-colorPrimary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-11 MuiSwitch-input"
            id="switch"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiSwitch-track"
      />
    </span>
    <button
      class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-1 MuiIconButton-colorPrimary"
      id="add-reason"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </span>
    </button>
  </div>
  <ul
    class="MuiList-root MuiList-padding"
  >
    <li
      class="MuiListItem-root MuiListItem-dense"
    >
      <div>
        input 1
      </div>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense"
    >
      <div>
        input 2
      </div>
    </li>
  </ul>
</div>
`;

exports[`ReasonSection renders the section with two input and remove icon 1`] = `
<div
  class="MuiBox-root MuiBox-root-14"
>
  <div
    class="MuiBox-root MuiBox-root-15"
  >
    <h4
      class="MuiTypography-root MuiTypography-h6"
    >
      持有逻辑(至少3条)
    </h4>
    <span
      class="MuiSwitch-root makeStyles-root-4 MuiSwitch-sizeSmall makeStyles-sizeSmall-3"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiSwitch-switchBase makeStyles-switchBase-5 MuiSwitch-colorPrimary PrivateSwitchBase-checked-9 Mui-checked"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input-11 MuiSwitch-input"
            id="switch"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiSwitch-track"
      />
    </span>
    <button
      class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-1 MuiIconButton-colorPrimary"
      id="add-reason"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </span>
    </button>
  </div>
  <ul
    class="MuiList-root MuiList-padding"
  >
    <li
      class="MuiListItem-root MuiListItem-dense"
    >
      <div>
        input 1
      </div>
      <button
        class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-1 MuiIconButton-colorSecondary"
        id="remove-0"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10H7v-2h10v2z"
            />
          </svg>
        </span>
      </button>
    </li>
    <li
      class="MuiListItem-root MuiListItem-dense"
    >
      <div>
        input 2
      </div>
      <button
        class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-1 MuiIconButton-colorSecondary"
        id="remove-1"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10H7v-2h10v2z"
            />
          </svg>
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`ReasonSection renders the section without any inputs 1`] = `
<div
  class="MuiBox-root MuiBox-root-6"
>
  <div
    class="MuiBox-root MuiBox-root-7"
  >
    <h4
      class="MuiTypography-root MuiTypography-h6"
    >
      持有逻辑(至少3条)
    </h4>
    <span
      class="MuiSwitch-root makeStyles-root-4 MuiSwitch-sizeSmall makeStyles-sizeSmall-3"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-8 MuiSwitch-switchBase makeStyles-switchBase-5 MuiSwitch-colorPrimary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-11 MuiSwitch-input"
            id="switch"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSecondary"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
        </span>
      </span>
      <span
        class="MuiSwitch-track"
      />
    </span>
    <button
      class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton-1 MuiIconButton-colorPrimary"
      id="add-reason"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm-2 10h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
          />
        </svg>
      </span>
    </button>
  </div>
  <ul
    class="MuiList-root MuiList-padding"
  />
</div>
`;
